In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('/content/AirQualityUCI[1].csv', sep=';', decimal = ',')

# **1.) Loading the Data and Understanding it and Simple Data Preparation**

In [ ]:
df.head()

,Date,Time,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Unnamed: 15,Unnamed: 16
0,10/03/2004,18.00.00,2.6,1360.0,150.0,11.9,1046.0,166.0,1056.0,113.0,1692.0,1268.0,13.6,48.9,0.7578,NaN,NaN
1,10/03/2004,19.00.00,2.0,1292.0,112.0,9.4,955.0,103.0,1174.0,92.0,1559.0,972.0,13.3,47.7,0.7255,NaN,NaN
2,10/03/2004,20.00.00,2.2,1402.0,88.0,9.0,939.0,131.0,1140.0,114.0,1555.0,1074.0,11.9,54.0,0.7502,NaN,NaN
3,10/03/2004,21.00.00,2.2,1376.0,80.0,9.2,948.0,172.0,1092.0,122.0,1584.0,1203.0,11.0,60.0,0.7867,NaN,NaN
4,10/03/2004,22.00.00,1.6,1272.0,51.0,6.5,836.0,131.0,1205.0,116.0,1490.0,1110.0,11.2,59.6,0.7888,NaN,NaN


In [ ]:
df.describe()

,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Unnamed: 15,Unnamed: 16
count,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,0.0,0.0
mean,-34.207524,1048.990061,-159.090093,1.865683,894.595276,168.616971,794.990168,58.148873,1391.479641,975.072032,9.778305,39.485380,-6.837604,NaN,NaN
std,77.657170,329.832710,139.789093,41.380206,342.333252,257.433866,321.993552,126.940455,467.210125,456.938184,43.203623,51.216145,38.976670,NaN,NaN
min,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,NaN,NaN
25%,0.600000,921.000000,-200.000000,4.000000,711.000000,50.000000,637.000000,53.000000,1185.000000,700.000000,10.900000,34.100000,0.692300,NaN,NaN
50%,1.500000,1053.000000,-200.000000,7.900000,895.000000,141.000000,794.000000,96.000000,1446.000000,942.000000,17.200000,48.600000,0.976800,NaN,NaN
75%,2.600000,1221.000000,-200.000000,13.600000,1105.000000,284.000000,960.000000,133.000000,1662.000000,1255.000000,24.100000,61.900000,1.296200,NaN,NaN
max,11.900000,2040.000000,1189.000000,63.700000,2214.000000,1479.000000,2683.000000,340.000000,2775.000000,2523.000000,44.600000,88.700000,2.231000,NaN,NaN


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(15), object(2)
memory usage: 1.2+ MB


In [ ]:
#we are seeing some Unnamed columns so lets drop them and drop the  empty rows as well if any are present
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]
df = df.dropna(how='all')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


In [ ]:
print("Shape", df.shape)

Shape (9357, 15)


In [ ]:
# as it was mentioned in the doc the Null values are marked as '-200' in the excel , lets see how many null values are present for every columns so it helps us to
# make imputing decisions on the columns

total_missing_count = (df == -200).sum()
print(total_missing_count)
print()
print("Percentage")
per_missing_count_col = ( total_missing_count / len(df) ) * 100
print(per_missing_count_col)

Date                0
Time                0
CO(GT)           1683
PT08.S1(CO)       366
NMHC(GT)         8443
C6H6(GT)          366
PT08.S2(NMHC)     366
NOx(GT)          1639
PT08.S3(NOx)      366
NO2(GT)          1642
PT08.S4(NO2)      366
PT08.S5(O3)       366
T                 366
RH                366
AH                366
dtype: int64

Percentage
Date              0.000000
Time              0.000000
CO(GT)           17.986534
PT08.S1(CO)       3.911510
NMHC(GT)         90.231912
C6H6(GT)          3.911510
PT08.S2(NMHC)     3.911510
NOx(GT)          17.516298
PT08.S3(NOx)      3.911510
NO2(GT)          17.548360
PT08.S4(NO2)      3.911510
PT08.S5(O3)       3.911510
T                 3.911510
RH                3.911510
AH                3.911510
dtype: float64


So as we can see from the above findings that there are more missing values for NHMC which is almost 90% and we cannot even think of imputing this columns we have to drop this and also GT values are recorded from another sensor so we cannot consider these values to predict the Target varaible C6H6(GT). remaining columns have missing values around  3% we can use SImpleImputer or KNN imputer for these missing columns.

Also we have noticed other thing that some columns T, RH, AH have exactly
 3.911510% missing. This is not a coincidence — it means the sensor device itself went offline for certain hours, causing all sensor readings to go missing simultaneously. These are the same 366 rows across all these columns.


 Lets drop NMHC(GT)

 CO(GT), NOx(GT), NO2(GT) — impute with median:
~18% missing is significant but not fatal.

Sensor columns + T, RH, AH — impute with median:
Only 3.9% missing

C6H6(GT) — drop those rows with missing values: This is your target variable.


In [ ]:
#lets replace -200 with NaN
df = df.replace(-200, np.nan)

#drop the NMHC(GT)
df = df.drop(columns=['NMHC(GT)'])

#drop NaN values for C6H6 column

df = df.dropna(subset=['C6H6(GT)'])


print("Shape of the Data frame after dropping the cols and rows" , df.shape)

Shape of the Data frame after dropping the cols and rows (8991, 14)


In [ ]:
#lets fill the mssing values with Simple Imputer using median
# The reason to use Median Imputer is becasue , Mean is affected by the outliers lets say there is been industrial dispoal from factories into the one any gien daya,
# the pollutant values will become high and these drag mean towards higher side , but median dosent get affected so lets use Median

from sklearn.impute import SimpleImputer
cols_to_impute = ['CO(GT)', 'NOx(GT)', 'NO2(GT)',
                  'PT08.S1(CO)', 'PT08.S2(NMHC)', 'PT08.S3(NOx)',
                  'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH']

imputer = SimpleImputer(strategy='median')
df[cols_to_impute] = imputer.fit_transform(df[cols_to_impute])
print(df.isnull().sum())

Date             0
Time             0
CO(GT)           0
PT08.S1(CO)      0
C6H6(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               0
dtype: int64


In [ ]:
df.describe()

,CO(GT),PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH
count,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000
mean,2.069314,1099.833166,10.083105,939.153376,230.802135,835.493605,111.586142,1456.264598,1022.906128,18.317829,49.234201,1.025530
std,1.304487,217.080037,7.449820,266.831429,188.717210,256.817320,43.205808,346.206794,398.484288,8.832116,17.316892,0.403813
min,0.100000,647.000000,0.100000,383.000000,2.000000,322.000000,2.000000,551.000000,221.000000,-1.900000,9.200000,0.184700
25%,1.200000,937.000000,4.400000,734.500000,112.000000,658.000000,85.000000,1227.000000,731.500000,11.800000,35.800000,0.736800
50%,1.800000,1063.000000,8.200000,909.000000,178.000000,806.000000,109.000000,1463.000000,963.000000,17.800000,49.600000,0.995400
75%,2.600000,1231.000000,14.000000,1116.000000,280.000000,969.500000,132.000000,1674.000000,1273.500000,24.400000,62.500000,1.313700
max,11.900000,2040.000000,63.700000,2214.000000,1479.000000,2683.000000,333.000000,2775.000000,2523.000000,44.600000,88.700000,2.231000


Our dataset has Date and Time Columns and these really useful to predict the C6H6 pollutant, becasue there may be certian periods of hours where there may peak traffic on weekdays and certains days in a month where industries releases all the harmful gases into the air and these actaully give sudden spikes which are not the outliers , so lets do some feature Extraction on these two columns because these may give us certain patterns to us in predicting the Target varaible.

In [ ]:
# lets first combine the Date and Time column
df['DateTime'] = pd.to_datetime(df['Date'] + ' ' + df['Time'].str.replace('.', ':', regex=False) ,format='%d/%m/%Y %H:%M:%S')

In [ ]:
df['DateTime'].head()

,DateTime
0,2004-03-10 18:00:00
1,2004-03-10 19:00:00
2,2004-03-10 20:00:00
3,2004-03-10 21:00:00
4,2004-03-10 22:00:00


In [ ]:
# now using DateTime columns lets get Hour, DayOfWeek, Month as seperate columns
df['Hour']       = df['DateTime'].dt.hour
df['DayOfWeek']  = df['DateTime'].dt.dayofweek
df['Month']      = df['DateTime'].dt.month

In [ ]:
#Now lets drop the Actual DateTime Columns
df = df.drop(columns=['Date', 'Time', 'DateTime'])

# **2.) EDA**

# Univaraite Analysis for the columns to understand the distribution

In [ ]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [ ]:
numeric_cols = ['CO(GT)', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)',
                'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)',
                'PT08.S5(O3)', 'T', 'RH', 'AH']


fig = make_subplots(rows=4, cols=3, subplot_titles=numeric_cols)


for i, col in enumerate(numeric_cols):
    row = (i // 3) + 1
    col_pos = (i % 3) + 1

    mean_val = df[col].mean()
    median_val = df[col].median()

    fig.add_trace(go.Histogram(x = df[col], nbinsx=50, name=col, marker_color='steelblue',opacity=0.8),row=row, col=col_pos)
    fig.add_vline(x=median_val,line_dash='dash', line_color='green',annotation_text=f'Median: {median_val:.2f}',row=row, col=col_pos)
    fig.add_vline(x = mean_val,line_color='red',annotation_text=f'Mean: {mean_val:.2f}',row=row, col=col_pos )

fig.update_layout(height=1100, width=1100, title_text='Univariate Distributions — All Numeric Features')
fig.update_layout(showlegend=False)
fig.show()

**Observations from the Above Histogram chart**

**Right Skewed Columns**
NOx(GT)  most extreme. Most hours have moderate NOx levels but occasional heavy traffic/industrial events create a very long right tail.

CO(GT) Same pattern as NOx(GT)

C6H6(GT) Our target variable is right skewed. This is important — it means a plain linear regression may struggle since it assumes normally distributed errors.  this causes regression models to focus too much on outliers
We need to consider the Log Transformation or Consider Robust Regression it is less sensitive to the Extreme outliers.

**Normal Columns**
T, AH, RH these are seasonal cycles and are normally distributed


**Moderately Skewed Columns**

PT08.S1(CO), NO2(GT), PT08.S5(O3), PT08.S2(NMHC)

# Pollutant Trends Over Time

So it is important to see how  C6H6 is chaning hour by hour, day by day and month by month, becasue then we can see some patterns , how exactly Time and Day matters in predicting the Target varaible C6H6

Observations from the above three Line Graphs

Chart1 (Hour)

1.) The peaks are around 8-9am and 6-7pm
2.) The lowest point is early morning 5 clock

Chart 2 (Day of Week)
1.) Saturdays and Sundays have less C6H6
2.) Weekdays has peaks


Chart 3 (Month):

October has highest peak and August has Low peak and October, Nov and December has more pollutant compared to the Aug, Feb.


All three time features (Hour, DayOfWeek, Month) show clear
patterns in C6H6(GT), A model
without these features would be missing important learning patterns


In [ ]:
# lets do it for Hours first

hourly = df.groupby('Hour')['C6H6(GT)'].mean().reset_index()
fig1 = go.Figure()
fig1.add_trace(go.Scatter(x=hourly['Hour'], y=hourly['C6H6(GT)']))
fig1.update_layout(title=dict(text='Average C6H6(GT) by Hour', x=0.5),xaxis_title='Hour of Day',yaxis_title='Average of C6H6',height=400)
fig1.show()



day_map = {0:'Mon', 1:'Tue', 2:'Wed', 3:'Thu', 4:'Fri', 5:'Sat', 6:'Sun'}
daily = df.groupby('DayOfWeek')['C6H6(GT)'].mean().reset_index()
daily['DayName'] = daily['DayOfWeek'].map(day_map)

fig2 = go.Figure()

fig2.add_trace(go.Scatter(x=daily['DayName'], y=daily['C6H6(GT)']))
fig2.update_layout(title=dict(text='Average C6H6(GT) by Day', x=0.5),xaxis_title='DayName',yaxis_title='Average of C6H6',height=400)
fig2.show()



month_map = {1:'Jan', 2:'Feb', 3:'Mar', 4:'Apr', 5:'May', 6:'Jun',
             7:'Jul', 8:'Aug', 9:'Sep', 10:'Oct', 11:'Nov', 12:'Dec'}

monthly = df.groupby('Month')['C6H6(GT)'].mean().reset_index()
monthly['MonthName'] = monthly['Month'].map(month_map)

fig3 = go.Figure()

fig3.add_trace(go.Scatter(x=monthly['MonthName'], y=monthly['C6H6(GT)']))
fig3.update_layout(title=dict(text='Average C6H6(GT) by Month', x=0.5),xaxis_title='MonthName',yaxis_title='Average of C6H6',height=400)
fig3.show()

**HeatMap**

In [ ]:
import plotly.express as px

numeric_cols = ['CO(GT)', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)',
                'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)',
                'PT08.S5(O3)', 'T', 'RH', 'AH', 'Hour', 'DayOfWeek', 'Month']


corr_matrix = df[numeric_cols].corr().round(2)
print(corr_matrix)

               CO(GT)  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
CO(GT)           1.00         0.81      0.84           0.83     0.77   
PT08.S1(CO)      0.81         1.00      0.88           0.89     0.66   
C6H6(GT)         0.84         0.88      1.00           0.98     0.66   
PT08.S2(NMHC)    0.83         0.89      0.98           1.00     0.64   
NOx(GT)          0.77         0.66      0.66           0.64     1.00   
PT08.S3(NOx)    -0.64        -0.77     -0.74          -0.80    -0.60   
NO2(GT)          0.64         0.59      0.56           0.59     0.75   
PT08.S4(NO2)     0.57         0.68      0.77           0.78     0.21   
PT08.S5(O3)      0.79         0.90      0.87           0.88     0.74   
T                0.01         0.05      0.20           0.24    -0.26   
RH               0.04         0.11     -0.06          -0.09     0.19   
AH               0.02         0.14      0.17           0.19    -0.16   
Hour             0.33         0.33      0.34           0.39     

In [ ]:
fig = px.imshow(corr_matrix, text_auto=True, height=700, width=900)
fig.show()

**Observations**

Strong Correlations:-
1.   PT08.S1(CO)      0.88            CO sensor strongly tracks Benzene
2.   PT08.S2(NMHC)    0.98            Strongest predictor
3.   PT08.S5(O3)      0.87            O3 sensor strongly correlated
4.   CO(GT)           0.84            Actual CO concentration tracks Benzene well
5.   PT08.S4(NO2)     0.77            Moderate-strong

Multicollinearity :-
1. PT08.S2(NMHC) & PT08.S1(CO)
2. PT08.S2(NMHC) & PT08.S5(O3)
3. PT08.S1(CO) & PT08.S5(O3)
4. PT08.S3(NOx) & PT08.S2(NMHC)  this is highly negative almost -0.80


# Key Takeaways

1. PT08.S2(NMHC) has 0.98 correlation with C6H6(GT) —
   almost perfect linear relationship. This will likely
   dominate the model as the single strongest predictor.

2. Several sensor pairs are highly correlated with each
   other (above 0.88) — multicollinearity is present.

3.  Time features (Hour, DayOfWeek, Month) show weak
   correlations individually but the line charts earlier
   showed clear patterns

Lets do scatter Plots for the Top features ['PT08.S2(NMHC)', 'PT08.S1(CO)', 'PT08.S5(O3)', 'CO(GT)', 'PT08.S3(NOx)']

In [ ]:
top_features = ['PT08.S2(NMHC)', 'PT08.S1(CO)', 'PT08.S5(O3)',  'CO(GT)', 'PT08.S3(NOx)']

for feature in top_features:
  fig = px.scatter(df, x = feature, y = 'C6H6(GT)', trendline = 'ols', title=f'{feature} vs C6H6(GT) Scatter Plot')
  fig.show()

# Key Takeaways

1.  PT08.S2(NMHC) has a curved relationship with C6H6(GT)
2. Heteroscedasticity is present in PT08.S1(CO) and
   PT08.S5(O3)
3. PT08.S3(NOx) has a hyperbolic negative relationship


These findings directly motivate us to do feature engineering
step where we apply log transformations to skewed and
nonlinear features.


# Feature Engineering

Non Linear Transformations

1. C6H6 (GT) the target variable is completely right skewed and it means a plain linear regression may struggle since it assumes normally distributed errors. this causes regression models to focus too much on outliers We need to consider the Log Transformation

2. PT08.S2(NMHC) has a curved relationship so we need to apply tansformation for this as well

3. PT08.S3(NOx) is Skewed + hyperbolic scatter plot and this is a negative corrlation but still it helps us becasue When PT08.S3(NOx) is high, C6H5 tends to be low. When it is low, C6H5 tends to be high.




In [ ]:
# before lets observe the skewness for the columns before transformation
numeric_cols = ['CO(GT)', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)',
                'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)',
                'PT08.S5(O3)', 'T', 'RH', 'AH']

skewness = df[numeric_cols].skew()
print(skewness[abs(skewness) > 1])

CO(GT)          1.616442
C6H6(GT)        1.361532
NOx(GT)         1.995177
PT08.S3(NOx)    1.101729
dtype: float64


In [ ]:
df['log_PT08_S3_NOx'] = np.log1p(df['PT08.S3(NOx)'])
df['log_PT08_S1_CO'] = np.log1p(df['PT08.S1(CO)'])

df['sqrt_PT08_S5_O3']  = np.sqrt(df['PT08.S5(O3)'])

df['log_C6H6'] = np.log1p(df['C6H6(GT)'])

In [ ]:
# Cyclical Encoding for Hour and Month becasue  Raw numbers treat Hour 23 and Hour 0 as far apart

df['hour_sin'] = np.sin(2 * np.pi * df['Hour'] / 24)
df['hour_cos'] = np.cos(2 * np.pi * df['Hour'] / 24)

df['month_sin'] = np.sin(2 * np.pi * df['Month'] / 12)
df['month_cos'] = np.cos(2 * np.pi * df['Month'] / 12)

In [ ]:
cols_to_drop = ['PT08.S3(NOx)', 'PT08.S5(O3)', 'PT08.S1(CO)',
                'Hour', 'Month', 'C6H6(GT)']

df_model = df.drop(columns=cols_to_drop)

In [ ]:
print(df_model.columns.tolist())

['CO(GT)', 'PT08.S2(NMHC)', 'NOx(GT)', 'NO2(GT)', 'PT08.S4(NO2)', 'T', 'RH', 'AH', 'DayOfWeek', 'log_PT08_S3_NOx', 'log_PT08_S1_CO', 'sqrt_PT08_S5_O3', 'log_C6H6', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos']


In [ ]:
df_model.shape

(8991, 17)

#  TRAIN - TEST SPLIT

In [ ]:
from sklearn.model_selection import train_test_split

X = df_model.drop(columns=['log_C6H6'])
y = df_model['log_C6H6']

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (8991, 16)
Target shape: (8991,)


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 42, shuffle = True)

print("\nTrain size:", X_train.shape)
print("Test size :", X_test.shape)


Train size: (7192, 16)
Test size : (1799, 16)


#Scaling

**let use RobustScaler**

Even after log/sqrt transformations, some outliers remain

RobustScaler uses median and IQR — not affected by outliers

Why Scaling after splitting the data:-

if we apply Robust scaler before splitting the median and IQR values will be different and this actually affects the model training , information will get leaked from the testdata during the trainign itself

In [ ]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)


# Convert back to dataframe,  After fit_transform(), sklearn returns a numpy array — all column names are lost.
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns)
X_test_scaled  = pd.DataFrame(X_test_scaled,  columns=X_test.columns)


print("Scaling complete")
print("X_train_scaled shape:", X_train_scaled.shape)
print("X_test_scaled shape :", X_test_scaled.shape)

print("\nSample of scaled values:")
X_train_scaled.head(3)

Scaling complete
X_train_scaled shape: (7192, 16)
X_test_scaled shape : (1799, 16)

Sample of scaled values:


,CO(GT),PT08.S2(NMHC),NOx(GT),NO2(GT),PT08.S4(NO2),T,RH,AH,DayOfWeek,log_PT08_S3_NOx,log_PT08_S1_CO,sqrt_PT08_S5_O3,hour_sin,hour_cos,month_sin,month_cos
0,0.000000,-0.651285,0.000000,0.000000,-0.303855,1.158730,-0.869403,0.443555,0.50,0.826283,-0.994910,-1.175473,0.000000,-0.707107,-0.633975,-0.5
1,0.538462,0.564272,0.171598,0.234043,0.721088,1.095238,-0.951493,0.177596,-0.75,0.075508,-0.217223,0.186171,0.183013,-0.683013,0.000000,-1.0
2,-0.538462,-0.622281,-0.656805,-0.595745,-0.489796,-0.436508,-0.104478,-0.571752,0.75,1.148467,-0.328969,-0.886791,-0.500000,0.500000,0.732051,0.0


In [ ]:
X_train_scaled.shape

(7192, 16)

# **Baseline Model only  Multi Linear Regression without Regularization** not the OLS which shows statistics where P values and adjusted R^2 everyhting is shown

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error


baseline_model = LinearRegression()
baseline_model.fit(X_train_scaled, y_train)

y_pred_train = baseline_model.predict(X_train_scaled)
y_pred_test  = baseline_model.predict(X_test_scaled)


In [ ]:
def evalaute_model(y_true, y_pred, dataset_name):
  r2   = r2_score(y_true, y_pred)
  mae  = mean_absolute_error(y_true, y_pred)
  mse  = mean_squared_error(y_true, y_pred)
  rmse = np.sqrt(mean_squared_error(y_true, y_pred))
  mape = np.mean(np.abs((y_true - y_pred) / y_true)) * 100


  print(f"{dataset_name} Results:")
  print(f"R2 Score: {r2:.4f}")
  print(f"MAE: {mae:.4f}")
  print(f"MSE: {mse:.4f}")
  print(f"RMSE: {rmse:.4f}")
  print(f"MAPE: {mape:.2f}%")
  print()
  return r2, mae, mse, rmse, mape

In [ ]:
train_metrics = evalaute_model(y_train, y_pred_train, "TRAIN")
test_metrics  = evalaute_model(y_test,  y_pred_test,  "TEST")

TRAIN Results:
R2 Score: 0.9746
MAE: 0.0835
MSE: 0.0122
RMSE: 0.1105
MAPE: 5.38%

TEST Results:
R2 Score: 0.9726
MAE: 0.0868
MSE: 0.0129
RMSE: 0.1136
MAPE: 5.22%



#Observations from the results

1. Baseline linear regression achieves R² = 0.9726 on
   test data which is really good that too with No regularization

2. Train-Test gap of 0.002 indicates no overfitting, model generalizes very well

3. MAPE of 5.22%  which means model accurately predict the C6H5 values


4. I think most of the weightage is went for the PT08.S2
   (NMHC) sensor has 0.98 correlation with target this single feature drives most of the prediction , even if we apply Lasso and Ridge or elastic net we many not see much improvements, but they will tell us which features are important. lets check that



As we know already Lasso will drive weak coefficients to zero and Ridge will handle multicollinearity between correlated sensor columns

# REGULARIZATION

**Hyperparameter Tuning of Lasso and Ridge Regression using Cross-Validation (R² Evaluation Metric)**

In [ ]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.model_selection import cross_val_score


alphas = [0.001, 0.01, 0.1, 1, 10, 100, 1000]

# for finding the best alphas for Lasso and Ridge we need this cross validation
# it finds the best alpha systematically by testing multiple values


ridge_scores = {}
lasso_scores = {}

for al in alphas:
  ridge_model = Ridge(alpha=al)
  lasso_model = Lasso(alpha=al)
  ridge_scores[al] = cross_val_score(ridge_model, X_train_scaled, y_train, cv=5, scoring='r2')
  lasso_scores[al] = cross_val_score(lasso_model , X_train_scaled, y_train, cv=5, scoring='r2')
  print(f"Alpha: {al}")
  print(f"Ridge MSE scores: {ridge_scores[al]}")
  print(f"Lasso MSE scores: {lasso_scores[al]}")
  print()



Alpha: 0.001
Ridge MSE scores: [0.97477232 0.97465904 0.97385922 0.97185859 0.9753748 ]
Lasso MSE scores: [0.97397535 0.97309622 0.97288357 0.97266553 0.97403472]

Alpha: 0.01
Ridge MSE scores: [0.97477231 0.9746586  0.97385846 0.97186074 0.97537436]
Lasso MSE scores: [0.96634501 0.96578615 0.96682327 0.9651172  0.96674098]

Alpha: 0.1
Ridge MSE scores: [0.97477224 0.97465414 0.97385086 0.97188212 0.97536993]
Lasso MSE scores: [0.91394572 0.91343704 0.91565637 0.91993949 0.91551323]

Alpha: 1
Ridge MSE scores: [0.97476736 0.97460733 0.97377237 0.97208263 0.97532324]
Lasso MSE scores: [-1.59636873e-03 -4.54960287e-06 -1.24745497e-07 -2.20940151e-03
 -1.25030917e-05]

Alpha: 10
Ridge MSE scores: [0.97444382 0.97400191 0.97284979 0.9731609  0.97470973]
Lasso MSE scores: [-1.59636873e-03 -4.54960287e-06 -1.24745497e-07 -2.20940151e-03
 -1.25030917e-05]

Alpha: 100
Ridge MSE scores: [0.96916888 0.96870464 0.9656259  0.97015382 0.96914034]
Lasso MSE scores: [-1.59636873e-03 -4.54960287e-06 -

In [ ]:
# we can clearly see that 0.001 has best alpha values , we can still find the best values again in that range but I think there wont be uch improvements

# Train Ridge

ridge_model = Ridge(alpha=0.001)
ridge_model.fit(X_train_scaled, y_train)

print("RIDGE RESULTS:")
evalaute_model(y_train, ridge_model.predict(X_train_scaled), "Train")
evalaute_model(y_test,  ridge_model.predict(X_test_scaled),  "Test")

RIDGE RESULTS:
Train Results:
R2 Score: 0.9746
MAE: 0.0835
MSE: 0.0122
RMSE: 0.1105
MAPE: 5.38%

Test Results:
R2 Score: 0.9726
MAE: 0.0868
MSE: 0.0129
RMSE: 0.1136
MAPE: 5.22%



(0.9726401444519215,
 0.08684660667885626,
 0.012899075709293863,
 np.float64(0.11357409788016748),
 np.float64(5.222150727793127))

In [ ]:
# Train Lasso

lasso_model = Lasso(alpha=0.001)
lasso_model.fit(X_train_scaled, y_train)

print("LASSO RESULTS:")
evalaute_model(y_train, lasso_model.predict(X_train_scaled), "Train")
evalaute_model(y_test,  lasso_model.predict(X_test_scaled),  "Test")

LASSO RESULTS:
Train Results:
R2 Score: 0.9737
MAE: 0.0848
MSE: 0.0126
RMSE: 0.1123
MAPE: 5.56%

Test Results:
R2 Score: 0.9720
MAE: 0.0878
MSE: 0.0132
RMSE: 0.1150
MAPE: 5.35%



(0.9719636817276563,
 0.08775018810826547,
 0.01321800078108298,
 np.float64(0.11496956458595023),
 np.float64(5.3489694729334065))

#Observations

1. Ridge at alpha=0.001 produces identical results to
   baseline model

2. Lasso at alpha=0.001 shows little performance drop than baseline and Ridge model

3. Baseline linear regression is already near optimal

4. Lasso will bring some coeffeicents to zero , SO lets look at the feature Seelction - then we can see which coefficients it took to zero. This will also help how genuinely all the features are important


In [ ]:
print("LASSO COEFFICIENTS:")
print()

print("feature               Coefficient")

for feature, coef in zip(X_train_scaled.columns, lasso_model.coef_):
    print(f"{feature:<20} {coef:>12.6f}")

LASSO COEFFICIENTS:

feature               Coefficient
CO(GT)                  -0.048498
PT08.S2(NMHC)            0.834518
NOx(GT)                 -0.034090
NO2(GT)                  0.026583
PT08.S4(NO2)            -0.044538
T                        0.037118
RH                       0.036423
AH                      -0.000000
DayOfWeek                0.017714
log_PT08_S3_NOx         -0.044561
log_PT08_S1_CO           0.000000
sqrt_PT08_S5_O3          0.169213
hour_sin                -0.107513
hour_cos                -0.060081
month_sin               -0.023756
month_cos               -0.052523


#Observations & these observations are also used for feature selection

1.  PT08.S2(NMHC) dominates with coefficient 0.835, meaning  when hydrocarbon levels rise, Benzene rises sharply.

2. sqrt_PT08_S5_O3 → +0.169 Second strongest. O3 sensor positively tracks Benzene, When traffic increases, both Benzene and Ozone precursors rise together.

3. hour_sin → -0.108 Third strongest ,  time of day has meaningful effect.
Negative coefficient on sine component means certain hours (late night/early morning) are associated with lower Benzene. Validates our datetime feature engineering.


4. hour_cos → -0.060 Second cyclical hour component,  rush hour peaks and overnight lows.


5. month_cos → -0.053
Seasonal effect present. Negative cosine component captures winter vs summer pollution patterns we saw in the monthly line chart.


6. CO(GT) → -0.048
Negative despite 0.84 correlation with Benzene. This is a multicollinearity effect — once PT08.S2(NMHC) accounts for most variance, CO(GT) compensates with slight negative coefficient to avoid double counting shared information.


7. log_PT08_S3_NOx → -0.045
Negative relationship consistent with scatter plot — as NOx sensor reading increases, Benzene decreases. Physically this could mean NOx dominates during diesel-heavy traffic while Benzene dominates during petrol-heavy traffic.


8. PT08.S4(NO2) → -0.045
Similar to NOx sensor


9. T → +0.037
Temperature mild positive effect. Warmer temperatures increase evaporation of fuel compounds including Benzene from vehicle fuel systems.


10. RH → +0.036
Relative humidity mild positive. Humidity affects how pollutants disperse and interact chemically in the atmosphere.


11. AH → 0.000, Eliminated.  redundant since T and RH are already in model.

12. log_PT08_S1_CO → 0.000   CO sensor information is already captured by CO(GT) and PT08.S2(NMHC) — completely redundant.



#Let’s use OLS as well to examine the p-values for feature selection and other evaluation metrics like adjusted R²

However, the problem with p-values is that they come from a statistical framework and p-values are affected by multicollinearity. Because of this, we may sometimes see high p-values (> 0.05) for features that are actually important.

On the other hand, Lasso focuses purely on prediction error + regularization.

Additionally, we cannot rely solely on R² or adjusted R². We also need to perform residual analysis and examine diagnostic plots before concluding whether linear regression is suitable for the dataset.

We doing OLS for its its model summary mainly for feature selection by examining the p-values of the features, and we also consider Lasso coefficients for feature selection. This helps us observe how different these approaches are in practice.

However, when it comes to selecting features for prediction, domain knowledge is extremely important. We cannot rely completely on p-values or Lasso coefficients alone.

In [ ]:
X_train_scaled = X_train_scaled.reset_index(drop=True)
X_test_scaled  = X_test_scaled.reset_index(drop=True)
y_train        = y_train.reset_index(drop=True)
y_test         = y_test.reset_index(drop=True)

# robust scaling or any operation that returns a new DataFrame and indices can get misaligned and this affects OLS but SKLearn models will not get affected by index missalignment

In [ ]:
import statsmodels.api as sm

X_train_sm = sm.add_constant(X_train_scaled)
X_test_sm  = sm.add_constant(X_test_scaled)

ols_model = sm.OLS(y_train, X_train_sm).fit()
print(ols_model.summary())

                            OLS Regression Results                            
Dep. Variable:               log_C6H6   R-squared:                       0.975
Model:                            OLS   Adj. R-squared:                  0.974
Method:                 Least Squares   F-statistic:                 1.718e+04
Date:                Mon, 23 Mar 2026   Prob (F-statistic):               0.00
Time:                        14:39:01   Log-Likelihood:                 5637.4
No. Observations:                7192   AIC:                        -1.124e+04
Df Residuals:                    7175   BIC:                        -1.112e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               2.1083      0.002   11

# I didn’t expect such a large difference between the p-values and the Lasso coefficients. The p-values suggest that almost every feature is important, since most of them are close to 0.00, except for two feature, log_PT08_S1_CO and AH. Interestingly, this same features are also eliminated by Lasso, as there coefficients were shrunk to zero.

However, when I compare these results with the domain knowledge I have, the feature importance suggested by Lasso seems more reasonable than what the p-values indicate. This makes me feel that Lasso has done a better job in identifying the truly relevant features.

**So lets rely on the Lasso values only and ofcourse the coeff values in the OLS model summary is also kind of similar to the Lasso coeff**

#Residual Analysis

1. Residuals vs Fitted Plot — checks heteroscedasticity or curves

2.  Q-Q Plot — checks normality more precisely

# Residuals vs Fitted Plot

In [ ]:
y_pred_train_lasso = lasso_model.predict(X_train_scaled)
y_pred_test_lasso  = lasso_model.predict(X_test_scaled)

residuals_train = y_train - y_pred_train_lasso
residuals_test  = y_test  - y_pred_test_lasso

In [ ]:
fig1 = px.scatter(
    x=y_pred_test_lasso,
    y=residuals_test,
    opacity=0.5,
    title='Residuals vs Fitted Values',
    labels={'x': 'Fitted Values (log scale)',
            'y': 'Residuals'}
)
fig1.add_hline(y=0, line_dash='dash', line_color='red', line_width=2)
fig1.update_layout(title_x=0.5, height=450)
fig1.show()

# Observations from the Residuals vs Fitted Plot  .

So generally here we observe three common patterns when we are doing this plot

1. The Random Scatter , where the points are scattered randomly and this means every thing is good

2. If we observe a U-Shape or the Curve it means relationship is not linear so to fix this we need to use Ploynomial regression or Log Transform which we did but still we are seeing curved so we need to use Polynomial regression , and this one observation here

3. Funnel shape , if we a funnel shapre that means we are seeing heteroscedasticity which mean Rich predictions has more error and we are observing that in our residual analysis


**Outcome from this step is we need to use Polynomial regression we still need to find whoich degree we need to use, we sill figure it out in next steps**


# But what does this plot say abour our predictions

1. Overpredicts  at low Benzene hours  (early morning, clean air)

2. Underpredicts at medium Benzene hours (normal traffic)

3. Overpredicts  at high Benzene hours  (extreme pollution events)

what is the reason for above three observations

for 1st and 3rd point the reason is in the range of 0.5 to 1.5 and 2.5+ the residuals are mostly negative

for 2nd point we see the residuals are mostly positive



# Q-Q Plot

So there is a approach to do  Q-Q plot

1. First Step we need to take the residuals

2. Sort the residuals from smallest to largest

3. Assign probability to each sorted value

4. Find the theorectical normal quantiles we use this as a baseline to compare our residuals

5. plot the residuals vs theoretical normla qunatiles

6. draw the reference line

In [ ]:
import scipy.stats as stats
residuals  = (y_test  - y_pred_test_lasso).reset_index(drop=True)


qq = stats.probplot(residuals, dist='norm')
# print(qq)
theoretical_quantiles = qq[0][0]
sample_quantiles = qq[0][1]


ref_x = [theoretical_quantiles[0], theoretical_quantiles[-1]]
ref_y = [sample_quantiles[0], sample_quantiles[-1]]

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=theoretical_quantiles,
    y=sample_quantiles,
    mode='markers',
    marker=dict(color='steelblue', size=4, opacity=0.6),
    name='Residual Quantiles'
))


fig.add_trace(go.Scatter(
    x=ref_x,
    y=ref_y,
    mode='lines',
    line=dict(color='red', width=2, dash='dash'),
    name='Normal Reference Line'
))


fig.update_layout(
    title=dict(
        text='Q-Q Plot — Lasso Residuals vs Normal Distribution',
        x=0.5
    ),
    xaxis_title='Theoretical Quantiles (Perfect Normal Distribution)',
    yaxis_title='Sample Quantiles (Our Actual Residuals)',
    height=500,
    width=750
)

fig.show()

Observations from the Q-Q plot

1. There is a clear deviation from the straight line
, this means resiudlas are not normally distributed

2. I can also see a  S shape pattern not clearly but little S pattern which means Non Normal Distribution

What this means in simple terms is that the model is missing something may be important patterns or non-linearity

While performing the Q-Q plot analysis, I noticed something
important — we transformed the entire dataset during feature
engineering, including the target column. This means we are
comparing transformed predictions against transformed test
values, which affects our residual plots and evaluation metrics.

This raised a question — if we applied all these transformations,
why are we still observing these patterns in the residual plots?
The answer is that the metrics and plots look better than reality
because everything is in log scale, not in original units.

Since the end user always needs predictions in proper units,
we need to apply inverse transformations to get back to the
original µg/m³ scale and see the true evaluation metrics
and residuals.

In [ ]:
# ── Inverse transform ─────────────────────────────────────────
y_pred_log    = lasso_model.predict(X_test_scaled)
y_pred_actual = np.expm1(y_pred_log)
y_actual      = np.expm1(y_test)

print("METRICS IN LOG SCALE:")
evalaute_model(y_test, y_pred_log, "Test")

print("\nMETRICS IN ORIGINAL SCALE (µg/m³):")
evalaute_model(y_actual, y_pred_actual, "Test")

METRICS IN LOG SCALE:
Test Results:
R2 Score: 0.9720
MAE: 0.0878
MSE: 0.0132
RMSE: 0.1150
MAPE: 5.35%


METRICS IN ORIGINAL SCALE (µg/m³):
Test Results:
R2 Score: 0.8662
MAE: 1.1554
MSE: 7.6881
RMSE: 2.7727
MAPE: 11.62%



(0.866205756019868,
 1.1553837959418782,
 7.688107773258515,
 np.float64(2.7727437265745487),
 np.float64(11.62269793378465))

# Keytakeways and Observations


1. In log scale the model appears exceptional — R² of 0.9720
and MAPE of 5.35% suggest nearly perfect predictions.
However this is misleading because log transformation
compresses the variance, making errors appear smaller
than they actually are.


2. In original scale the true picture emerges — R² drops
to 0.8662 meaning the model explains only 86.62% of
variance


3. MAPE of 11.62% in original scale is the most honest
metric — on average our predictions deviate by 11.62%
from actual Benzene values. This is more than double
the log scale MAPE of 5.35%, confirming that log
transformation was masking

# Key Takeaway was always report metrics in original units for interpretability.

Again as you we cannot completely rely on the Evalaution metrics and log transformations lets see the Residual analysis and come to a final conclusion.

In [ ]:
# Residuals in original scale
residuals_actual = y_actual - y_pred_actual

# Residuals vs Fitted — Original Scale
fig = px.scatter(
    x=y_pred_actual,
    y=residuals_actual,
    opacity=0.5,
    title='Residuals vs Fitted Values (Original Scale µg/m³)',
    labels={
        'x': 'Fitted Values (µg/m³)',
        'y': 'Residuals (µg/m³)'
    }
)
fig.add_hline(y=0, line_dash='dash', line_color='red', line_width=2)
fig.update_layout(title_x=0.5, height=450)
fig.show()

# lot of observations when I saw the plot after Inverse Log

1. Model over predicted for exteme benze values , which means the Lasso Multiple Linear model cannot be trusted for high values Benezen C6H5 predictions at all.


2. we are slightly seeing the Funnel shape which mean Heteroscedasticity


we can see after 21 x - axis values the spread got increased and also from here we are seeing a small funnel that means , for medium Beneze Target feature the model slighlty over predicts and for above 40 Target variable our Lasso model highly over predicts.

The root cause  is clearly Non Linear, we need to use Polynomial Multi Linear regression Model for this dataset

Also our maximum value in the C6H6 Target value was only  63.7 , here there are few points 70 so this also proves model is overpredicting

Lets try to imppement the Polynomial Multi Linear Regression + Lasso and observe how the model performs on the actual Target value Scale.

Lets use the Actual features that we have before scaling

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LassoCV

#lets take a degree of 2 to create polynomial features

poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)


In [ ]:
# now lets do scaling on polynomail features

X_train_poly.shape

(7192, 152)

In [ ]:
X_test_poly.shape

(1799, 152)

In [ ]:
scaler = RobustScaler()

X_train_poly_scaled = scaler.fit_transform(X_train_poly)
X_test_poly_scaled = scaler.transform(X_test_poly)

In [ ]:
lasso = LassoCV(cv=5, random_state=42)

lasso.fit(X_train_poly_scaled, y_train)
print("\nBest alpha choosed:", lasso.alpha_)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.3052475461843436, tolerance: 0.27261503696603956

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.3346680923003795, tolerance: 0.275585043369486

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.36675235494424463, tolerance: 0.275585043369486




Best alpha choosed: 0.0008591028614858096


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.591e+00, tolerance: 3.451e-01



In [ ]:
y_pred_log = lasso.predict(X_test_poly_scaled)

y_pred_actual = np.expm1(y_pred_log)
y_actual      = np.expm1(y_test)

print("\nMETRICS IN ORIGINAL SCALE (µg/m³):")
evalaute_model(y_actual, y_pred_actual, "Test")


METRICS IN ORIGINAL SCALE (µg/m³):
Test Results:
R2 Score: 0.9942
MAE: 0.2911
MSE: 0.3321
RMSE: 0.5763
MAPE: 3.60%



(0.9942197174354849,
 0.2911089299688181,
 0.33214758717481446,
 np.float64(0.5763224680461577),
 np.float64(3.5957147794631292))

In [ ]:
# Residuals in original scale
residuals_actual = y_actual - y_pred_actual

# Residuals vs Fitted — Original Scale
fig = px.scatter(
    x=y_pred_actual,
    y=residuals_actual,
    opacity=0.5,
    title='Residuals vs Fitted Values (Original Scale µg/m³) POLYNOMIAL LINEAR REGRESSION degree 2',
    labels={
        'x': 'Fitted Values (µg/m³)',
        'y': 'Residuals (µg/m³)'
    }
)
fig.add_hline(y=0, line_dash='dash', line_color='red', line_width=2)
fig.update_layout(title_x=0.5, height=450)
fig.show()

#Observations from the Degree 2 Poynomial Linear regression + Lasso on the Inverse Transformation Predictions

1. It performed way better than Multi Linear + Lasso Regression both in terms of Evaluation Metrics and also Residual analysis [Fitted vs Residuals]

2. Dramatic improvement over linear model —
   residual range compressed from (-55, +5) to
   (-4, +9) µg/m³, an 80% reduction in error magnitude.

3. residuals now randomly scattered above and below zero line,
   confirming polynomial model captures the curved
   relationship that linear model could not.

4. Mild heteroscedasticity remains

5. Model explains 99.42% of variance in actual Benzene, Exceptional performance in original units

6. Predictions within 3.6% of actual values Previously 11.62% — 3x improvement







# Final Conclusions - Air Quality Benzene Prediction [C6H6(GT)] Target

1. **PROBLEM DEFINITION**
   The goal was to predict Benzene concentration (C6H6(GT))
   in µg/m³ using sensor readings and environmental features
   from an Italian roadside air quality monitoring device.


2. **DATA UNDERSTANDING**
   Dataset contained 9358 hourly readings from March 2004
   to February 2005. Missing values were encoded as -200
   rather than NaN. Two separate devices collected data —
   cheap metal oxide sensors (PT08.S1-S5) and an expensive
   certified reference analyzer (GT columns).


3.  **MISSING VALUE TREATMENT**
   After replacing -200 with NaN, NMHC(GT) was dropped
   entirely due to 90.2% missingness. Target variable rows
   with missing C6H6(GT) were dropped since imputing a
   target variable would mean fabricating answers. All
   remaining columns were imputed using median rather than
   mean because pollution data is right skewed — median
   is resistant to extreme pollution spikes that would
   distort the mean.


4. **DATETIME FEATURE ENGINEERING**
   Date and Time columns were converted into Hour,
   DayOfWeek and Month features. Line charts confirmed
   these features capture meaningful patterns


6.  **EXPLORATORY DATA ANALYSIS**
   Univariate analysis revealed four highly skewed columns
   (NOx(GT) skewness 1.995, CO(GT) 1.616, C6H6(GT) 1.362,
   PT08.S3(NOx) 1.102)

7. **NONLINEAR TRANSFORMATIONS**
   Log transformation was applied to PT08.S3(NOx) reducing
   skewness from 1.102 to 0.027 and PT08.S1(CO) reducing
   skewness from 0.756. Square root was applied to
   PT08.S5(O3) reducing skewness from 0.628 to 0.164.
   Target C6H6(GT) was log transformed reducing skewness
   from 1.362 to -0.233.


8. **TRAIN TEST SPLIT AND SCALING**
   Data was split 80/20

9. **BASELINE LINEAR REGRESSION**
   OLS linear regression achieved R² = 0.9746 on training
   and 0.9726 on test in log scale with negligible gap of
   0.002 indicating no overfitting. However inverse
   transforming to original µg/m³ revealed true performance
   of R² = 0.8662.

   Log scale metrics were masking the real
   prediction difficulty highlighting why original scale
   evaluation is essential.


10. **REGULARIZATION**
   Ridge and Lasso were applied using cross validation
   to find optimal alpha. Both converged to alpha = 0.001
   indicating minimal regularization was needed — the
   dataset is large enough and clean enough that the
   baseline model was already near optimal. Ridge produced
   identical results to baseline while Lasso marginally
   reduced performance to R² = 0.9720 due to coefficient
   shrinkage.


11. **FEATURE IMPORTANCE**
    Lasso coefficient analysis identified PT08.S2(NMHC)
    as the dominant predictor with coefficient 0.835

    Two features were eliminated — AH (absolute humidity) due to
    redundancy with T and RH already in the model, and
    log_PT08_S1_CO due to overlap with other correlated
    sensor features.


12. **RESIDUAL ANALYSIS — LOG SCALE**
    Residuals vs fitted plot revealed mild curved pattern
    (inverted U shape) confirming remaining nonlinearity

13. **RESIDUAL ANALYSIS — ORIGINAL SCALE**
    Inverse transforming to µg/m³ revealed the true
    severity of model limitations hidden in log scale.
    Severe heteroscedasticity with residuals ranging from
    -55 to +5 µg/m³, at high values showing model overpredicting

14. **POLYNOMIAL REGRESSION — DEGREE 2**
    Applying degree 2 polynomial features transformed
    model performance dramatically in original scale —
    R² improved from 0.8662 to 0.9942, MAE reduced 75%
    from 1.1554 to 0.2911 µg/m³, RMSE reduced 79% from
    2.7727 to 0.5763 µg/m³ and MAPE improved from 11.62%
    to 3.60%. Residual plots confirmed systematic negative
    bias eliminated, impossible predictions removed and
    residual range compressed from 60 to 13 µg/m³.
    This validates that residual analysis correctly
    diagnosed the nonlinearity problem.

 15. **FINAL RECOMMENDATION**
    Polynomial regression degree 2 is the recommended
    model for Benzene prediction with R² = 0.9942 and
    MAE = 0.2911 µg/m³ in original units. Average
    prediction error of 0.29 µg/m³ .